go-controller/pkg/ovn/address_set.go

```diff
@@ -0,0 +1,272 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	"k8s.io/klog"
+)
+
+type AddressSetIterFunc func(hashedName, namespace, suffix string)
+type AddressSetDoFunc func(as AddressSet) error
+
+// AddressSetFactory is an interface for managing address set objects
+type AddressSetFactory interface {
+	// NewAddressSet returns a new object that implements NewAddressSet
+	// and contains the given IPs, or an error
+	NewAddressSet(name string, ips []net.IP) (AddressSet, error)
+	// ForEachAddressSet calls the given function for each address set
+	// known to the factory
+	ForEachAddressSet(iteratorFn AddressSetIterFunc) error
+}
+
+type ovnAddressSetFactory struct {
+	sync.Mutex
+	sets map[string]*ovnAddressSet
+}
+
+// NewOvnAddressSetFactory creates a new AddressSetFactory backed by
+// address set objects that execute OVN commands
+func NewOvnAddressSetFactory() AddressSetFactory {
+	return &ovnAddressSetFactory{
+		sets: make(map[string]*ovnAddressSet),
+	}
+}
+
+// ovnAddressSetFactory implements the AddressSetFactory interface
+var _ AddressSetFactory = &ovnAddressSetFactory{}
+
+// NewAddressSet returns a new address set object
+func (asf *ovnAddressSetFactory) NewAddressSet(name string, ips []net.IP) (AddressSet, error) {
+	asf.Lock()
+	defer asf.Unlock()
+	if _, ok := asf.sets[name]; ok {
+		klog.Errorf("address set %q already exists; overwriting", name)
+	}
+	as, err := newOvnAddressSet(name, ips, asf.removeAddressSet)
+	if err == nil {
+		asf.sets[name] = as
+	}
+	return as, err
+}
+
+// ForEachAddressSet will pass the hashedName, namespaceName and
+// the first suffix in the name to the 'iteratorFn' for every address_set in
+// OVN. (Each unhashed name for an ovnAddressSet can be of the form
+// namespaceName.suffix1.suffix2. .suffixN)
+func (asf *ovnAddressSetFactory) ForEachAddressSet(iteratorFn AddressSetIterFunc) error {
+	output, stderr, err := util.RunOVNNbctl("--format=csv", "--data=bare", "--no-heading",
```

Comment:
@trozet this is used for Sync() from namespace and policy to clean up left-overs when the master starts. We don't yet have them in the 'sets' field because nothing has added them there yet.

Reply:
oh, right...it seems a little awkward here with the refactor. I guess it could either be in the sync function itself, since you would never use this anywhere else. Not a huge deal because this will go away anyway when we get rid of all the sync functions in the future.